File: new/python/143_reorder_list_reverse_and_merge.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # idea: reverse and combine
        # break the linked list into two halves
        # reverse the second half of the linked list
        # combine first node from first linked list to first node from second linked list
        # to second node from first linked list to second node from second linked list ...

        # find the mid point to break the linked list
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # now slow points one before the middle node
        second = slow.next
        slow.next = None  # break the linked list into first and second linked list
        prev = None  # the prev pointer is going to be (later) the head of the second ll

        # reverse the second linked list
        while second:
            temp = second.next
            second.next = prev
            prev = second
            second = temp

        # combine the first and second linked list, now prev is the head of the second ll
        first = head
        second = prev

        while second:
            temp1 = first.next
            temp2 = second.next
            first.next = second
            second.next = temp1
            first = temp1
            second = temp2

    # time: O(n)
    # space: O(1)

File: new/cpp/143_reorder_list_reverse_and_merge.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        // idea: reverse and merge
        // break the linked list into two halves using slow and fast pointer to get the midpoint
        // after breaking the ll, reverse the second half of the linked list
        // after breaking and reversing, combine first node of first ll to first node of second ll
        // and first node of second ll to second node of first ll, and second node of first ll to second
        // node of second ll ...

        // find the midpoint
        ListNode* slow = head;
        ListNode* fast = head->next;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // now slow points to one before the start of the second linked list
        ListNode* second = slow->next;
        slow->next = nullptr; // now cut the first and second ll into two seperate ll

        // reverse the second ll
        ListNode* prev = nullptr;
        ListNode* temp;
        while (second != nullptr) {
            temp = second->next;
            second->next = prev;
            prev = second;
            second = temp;
        }

        // combine first ll and second ll alternatively
        ListNode* temp1;
        ListNode* temp2;
        ListNode* first = head;
        second = prev;
        while (second != nullptr) {
            temp1 = first->next;
            temp2 = second->next;
            first->next = second;
            second->next = temp1;
            first = temp1;
            second = temp2;
        }
    }
    // time: O(n)
    // space: O(1)
};
